### PR TITLE
Windows: Fix dragging and dropping files from compressed files into editor

### LINKED ISSUE #97128: Dragging and dropping files from 7zip into a Godot project fails.
### Tested versions

5d62e12fcbd46034dd06f91c2b20bdd366f3a0c9

### System information

Windows 11

### Issue description

Windows and third-party software specific issue:
Dragging and dropping files from an archive with 7zip currently does not work. It appears that 7zip will create a copy of the files being dragged from the archive in the Windows temp folder, and delete it once the action has finished. It seems the APIs being used to give drag and drop awareness on Windows are only responsive after 7zip deletes this temp file, resulting in an error.

While not a huge issue, its a minor usability snag and raises the question if there are alternative APIs we could or should be using on Win32 which are more compatible with the behaviour of software like 7zip. Certain other applications do appear to allow dragging files from 7zip on Windows, so it may be worth researching.

My interest in this stems from my usual concern of the beginner user experience; I imagine a lot of new developers might start learning by opening a zip file full of starter graphics/sounds/whatever, and it would be nice to make this process a bit more elegant.

### Steps to reproduce

- Create a new project
- Get any .zip file.
- Use 7zip to drag the contents of the .zip file into the Godot editor application.
- Observe the error which references the /temp/ path which not longer exists.

### Minimal reproduction project (MRP)

N/A

### PR DESCRIPTION
**Pull request by Garetonchick**
Fixes #97128.

Use case described by @Maran23 in https://github.com/godotengine/godot/issues/97128#issuecomment-2357090873 still doesn't work. I can look into it later.

### CODE REVIEW

File: platform/windows/drop_target_windows.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/**************************************************************************/
+/*  display_target_windows.h                                              */
```
Comment (RedMser) [inline platform/windows/drop_target_windows.h:2]: should be drop_target_windows.h

**Review by SaracenOne — APPROVED** [review]
Good job, and congratulations on your first commit! I tested this and it does appear to work in regards to the issue I raised with 7zip. I'll approve this for now, but would you mind squashing the commits from this PR into a single commit? We generally prefer that convention when merging.

**Comment by Garetonchick** [comment]
> Good job, and congratulations on your first commit! I tested this and it does appear to work in regards to the issue I raised with 7zip. I'll approve this for now, but would you mind squashing the commits from this PR into a single commit? We generally prefer that convention when merging.

Thank you! I have squashed the commits.

**Review by AThousandShips — APPROVED** [review]

File: platform/windows/drop_target_windows.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		int fcount = DragQueryFileW(hDropInfo, 0xFFFFFFFF, nullptr, 0);
+
+		for (int i = 0; i < fcount; i++) {
+			DragQueryFileW(hDropInfo, i, buf, buffsize);
+			String file = String::utf16((const char16_t *)buf);
+			files.push_back(file);
+		}
+
+		GlobalUnlock(stg.hGlobal);
+	}
```
Comment (bruvzg) [inline platform/windows/drop_target_windows.cpp:114]:
```suggestion
	Char16String buf;

	if (pDataObj->GetData(&fmt, &stg) != S_OK) {
		return E_UNEXPECTED;
	}

	HDROP hDropInfo = (HDROP)GlobalLock(stg.hGlobal);

	Vector<String> files;

	if (hDropInfo != nullptr) {
		int fcount = DragQueryFileW(hDropInfo, 0xFFFFFFFF, nullptr, 0);

		for (int i = 0; i < fcount; i++) {
			int buffsize = DragQueryFileW(hDropInfo, i, nullptr, 0);
			buf.resize(buffsize + 1);
			DragQueryFileW(hDropInfo, i, (LPWSTR)buf.ptrw(), buffsize + 1);
			String file = String::utf16((const char16_t *)buf.ptr());
			files.push_back(file);
		}

		GlobalUnlock(stg.hGlobal);
	}
```

Godot support long file names, so path length limit is 32767, not 4096.

**Review by bruvzg — APPROVED** [review]
Seems to be working as expected.

> Use case described by @Maran23 in https://github.com/godotengine/godot/issues/97128#issuecomment-2357090873 still doesn't work. I can look into it later.

This likely requires handling [CFSTR_FILECONTENTS](https://learn.microsoft.com/en-us/windows/win32/shell/clipboard?redirectedfrom=MSDN#cfstr_filecontents) and [CFSTR_FILEDESCRIPTOR](https://learn.microsoft.com/en-us/windows/win32/shell/clipboard?redirectedfrom=MSDN#cfstr_filedescriptor) in addition to `CF_HDROP`.

**Comment by Garetonchick** [comment]
> Seems to be working as expected.
> 
> > Use case described by @Maran23 in [#97128 (comment)](https://github.com/godotengine/godot/issues/97128#issuecomment-2357090873) still doesn't work. I can look into it later.
> 
> This likely requires handling [CFSTR_FILECONTENTS](https://learn.microsoft.com/en-us/windows/win32/shell/clipboard?redirectedfrom=MSDN#cfstr_filecontents) and [CFSTR_FILEDESCRIPTOR](https://learn.microsoft.com/en-us/windows/win32/shell/clipboard?redirectedfrom=MSDN#cfstr_filedescriptor) in addition to `CF_HDROP`.

Added handling of `CFSTR_FILEDESCRIPTOR` and `CFSTR_FILECONTENTS` as you have suggested. Now drag-and-drop from inside the `.zip` file from Windows Explorer also works. Also I tested the same case with Total Commander and it works as well.

However, this time I had to do a bit of filesystem manipulations and couldn't find some functinality in Godot's source code. These are: joining filepaths with correct OS separator, creating temporary directory with random name, creating all subdirectories for a path and recursively removing directory with it's contents. For now I have written quick & dirty implementations of these, which are not general purpose, but happen to work here. So I would appreciate it if someone pointed me to the relevant existing functionality. Or if it doesn't exist, it probably should be added (maybe in a separate PR?) for general use.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/44fa552343722bb048e2d7c6d3661174a95a8a3c), it works as expected on Windows 11 23H2. Nested file structures work correctly, and so do file names containing Unicode characters.

Note that empty folders are not created if you drag a folder that contains empty folders to the Godot editor. (Even if you right-click it and use **Show in File Manager**, the folders won't be there.)

As mentioned in OP, using 7-zip's GUI (`7ZFM.exe`) will work, but using Windows Explorer won't. This applies regardless of the compression format.

**Comment by bruvzg** [comment]
> These are: joining filepaths with correct OS separator, creating temporary directory with random name, creating all subdirectories for a path and recursively removing directory with it's contents. 

There's `DirAccess::erase_contents_recursive`, `DirAccess::make_dir_recursive` and `String::path_join` (like all other public Godot IO functions, these use `/` separator, usually conversion is done internally right before passing path to the system API and never exposed to the user, see `fix_path` in `OS_Windows`).

https://github.com/godotengine/godot/blob/b3bcb2dc14691f7729984128dca26a844f662fa1/platform/windows/os_windows.cpp#L93-L108

**Comment by Repiteo** [comment]
Do we want to wait and try to solve that edge-case, or is this ready to merge now in your eyes?

**Comment by Garetonchick** [comment]
> Do we want to wait and try to solve that edge-case, or is this ready to merge now in your eyes?

I think it's okay to merge. This edge case doesn't cause issues on my system (Windows 10), so I can't debug @Calinou's problem on Windows 11. I believe fixing it should be in a separate PR done by someone else.

**Comment by Repiteo** [comment]
Thanks! Congratulations on your first contribution! :tada:

**Pull request merged**